core/logger: add Critical* and Trace*

jmank88 opened the pull request:
https://app.shortcut.com/chainlinklabs/story/21618/introduce-more-granular-log-levels

- `Critical*`: `[CRIT]` (repurposing zap's `dpanic`)
- `Trace*`: `[DEBUG] [TRACE]` (iff `//go:build trace`)
- refactor some related Sentry and SQL log helpers

kaleofduty commented on core/logger/logger.go:
```diff
@@ -173,6 +188,18 @@ func (l *zapLogger) ErrorIfClosing(c io.Closer, name string) {
 }
 
+func (l *zapLogger) Critical(args ...interface{}) {
+	l.sugaredWithCallerSkip(1).DPanic(args...)
```
I'm not sure DPanic fits the semantics of what we're trying to do here:

https://github.com/uber-go/zap/blob/master/FAQ.md#whats-dpanic

jmank88 replied:
We are not actually using zap in development mode at all, so we are only re-purposing the existing level ordering without opting-in for development panics. But regardless, why doesn't this map to our use case? I thought this sounded like just what we wanted, particularly if we are going to start recovering from panics again:
> errors that are theoretically possible, but shouldn't actually happen, without crashing in production

kaleofduty replied:
Because as proposed critical would also be used for serious errors that are *expected to happen*

jmank88 replied:
Can you elaborate on what _expected_ means in this context? Any type of error is expected under the right conditions, but I would have thought that for a properly configured and maintained node we would not expect any Critical errors to be logged. Do we anticipate this being used as a regular notification mechanism, rather than limiting only to truly exceptional cases?

samsondav reviewed (commented):
I suspect implementing this might require us to dump zap

jmank88 commented:
> I suspect implementing this might require us to dump zap

This implementation should be fully functional in it's current form. The only compromise is that Trace logs are just Debug with prefixed messages. Is something else missing?

samsondav commented:
Yup this will work as-is just stating it would be nice to have actual logger support for TRACE/CRIT levels

cl-homu commented:
:umbrella: The latest upstream changes (presumably 73a70475700760b4f8f95097713062c4df3d0204) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jmank88 commented:
Rebased and refactored to accommodate the sentry changes. I think separating zap and having a generic wrapping `sentryLogger` is a bit cleaner and more flexible.

jmank88 commented on core/logger/sentry.go:
```diff
@@ -0,0 +1,325 @@
+	}
+}
+
+//TODO could include sentry event IDs with log lines: .With("sentryEvent",*EventID)
```
@samsondav did you consider including the returned `Capture*() EventID`s ? Would it be worthwhile for the node ops to have this ID for reference?

jmank88 replied:
Follow up story: https://app.shortcut.com/chainlinklabs/story/22266/sentry-log-eventids

jmank88 commented on core/logger/zap.go:
```diff
@@ -0,0 +1,114 @@
+	newLogger := *l
+	newLogger.name = joinName(l.name, name)
+	newLogger.SugaredLogger = l.SugaredLogger.Named(name)
+	newLogger.Trace("Named logger created")
```
I had already been thinking about adding a way to dump a map of these injected Named loggers anyways since it's a useful map of the system (like how some http router packages can print all the routes to generate docs).

jmank88 commented on core/recovery/recover.go:
```diff
@@ -21,7 +21,7 @@ func WrapRecover(lggr logger.Logger, fn func()) {
 			sentry.Flush(logger.SentryFlushDeadline)
 
-			lggr.Errorw("goroutine panicked", "panic", err, "stacktrace", string(debug.Stack()))
+			lggr.CriticalW("Recovered goroutine panic", "panic", err, "stacktrace", string(debug.Stack()))
```
Bumped these to critical level, and also noticed that this will double report via sentry above and then within the logger too. It could probably be worked around by doing something like (1) changing Critical* signatures to accept a proper `error` and (2) wrapping recovered panics in a `type Recovered` that the logger can check for and then use sentry `Recover()` instead.

jmank88 replied:
(Or perhaps just another set of three `Logger.Recovered*` methods would be cleaner :shrug: )

jmank88 replied:
Follow up story: https://app.shortcut.com/chainlinklabs/story/22268/sentry-wraprecover-double-reports

jmank88 commented on core/services/pg/q.go:
```diff
@@ -119,11 +119,15 @@ func NewQ(db *sqlx.DB, logger logger.Logger, config LogConfig, qopts ...QOpt) (q
 	}
 	q.db = db
-	q.logger = logger
+	q.logger = logger.Helper(2)
```
Skip two levels to point to the callers with the actual SQL instead of the inner helper methods. For example:
```
orm:67	err = q.Get(head, `SELECT * FROM heads WHERE evm_chain_id = $1 ORDER BY number DESC, created_at DESC, id DESC LIMIT 1`, orm.chainID)
```
instead of
```
pg:283	q.logger.Warnf("SLOW SQL QUERY: %d ms", elapsed.Milliseconds())
```

jmank88 commented on core/services/pg/q.go:
```diff
@@ -177,16 +181,30 @@ func (q Q) ExecQIter(query string, args ...interface{}) (sql.Result, context.Can
 		return errors.Wrap(err, "error binding arg")
 	}
-	return q.ExecQ(query, args...)
+	ctx, cancel := q.Context()
+	defer cancel()
+
+	q.logSqlQuery(query, args...)
+	begin := time.Now()
+	defer q.postSqlLog(ctx, begin)
+
+	_, err = q.Queryer.ExecContext(ctx, query, args...)
+	return q.withLogError(err)
```
Inlining these blocks so that the log helpers don't need to skip extra extra levels.

pinebit reviewed (approved):
LGTM!

jmank88 commented:
Merging to unblock a related story but can circle back for any additional feedback - and I created two follow-up stories as well (linked from comments).
